### PR TITLE
Async tests and fixes

### PR DESCRIPTION
**Pull request by camilbancioiu**
This PR adds the following changes:
* Change the timelock storage key prefix to `ARWEN@TIMELOCK`, taking advantage of the storage protection;
* Unexport the fields of the `asyncContext`;
* Add a guard against creating more than one `AsyncCall` to the same external contract (multiple calls to the same contract are currently unsupported);
* Remove gas redistribution after executing synchronous calls;
* Partially add accummulation of gas remaining into the `AsyncCall` - this will be finished after a bug is fixed;
* Simplify function `computeDataLengthFromArguments()` and fixes it, making it accurate.
* Change `finishSyncExecution()` to correctly set the `ReturnCode` of the `VMOutput` after the synchronous execution of a callback;
* Fix `createCallbackInput()` to correctly pass the `ReturnCode` of the destination call to the callback, as argument, even if the `ReturnCode` is `vmcommon.Ok`;
* Change `validator.verifyValidFunctionName()` to anforce proper and meaningful restrictions against function names (test updated);
* Move utility functions that manipulate `ContractCallInput` instances and `VMOutput` instances into `arwen/helpers.go` (previously found in `testInitializer`, but useful for tests of other packages as well);
* Improve `VmHostMock` and make it simulate execution, with enqueued `VMOutput` instances;
* Add two test contracts, written in pure Rust, but relying on `elrond-wasm-rs` for API wrappers; currently not used, but will be, after finishing implementing unit tests;
* Create `TestExecution_ExecuteOnDestContext_GasRemaining()`, which highlights a bug in `host.ExecuteOnDestContext()`, caused by calling `metering.GasLeft()` on the wrong Wasmer instance - fix pending.

### CODE REVIEW

File: integrationTests/json/vm_test.go
Comment (sasurobert) [inline integrationTests/json/vm_test.go:338]: still not working ?

File: mock/vmHostMock.go
Comment (sasurobert) [inline mock/vmHostMock.go:165]: GetNextVMOutput ?

File: arwen/contexts/async.go
Comment (sasurobert) [inline arwen/contexts/async.go:679]: why do we need this ?

Comment (sasurobert) [reply arwen/contexts/async.go:679]: and why do we need so exactly ?

Comment (sasurobert) [reply arwen/contexts/async.go:679]: I think the cost is already taken out when making asyncCall - there is a base cost plus data copy per byte, which should be enough.

File: arwen/contexts/async.go
Comment (sasurobert) [inline arwen/contexts/async.go:259]: in case of other errors ?

**Review by sasurobert — APPROVED** [review]

**Review by sasurobert — CHANGES_REQUESTED** [review]
need some small changes

File: arwen/helpers.go
Comment (BeniaminDrasovean) [inline arwen/helpers.go:276]: Missing comment.

File: arwen/helpers.go
Comment (BeniaminDrasovean) [inline arwen/helpers.go:291]: Missing comment.

File: arwen/helpers.go
Comment (BeniaminDrasovean) [inline arwen/helpers.go:304]: Missing comment.

File: arwen/helpers.go
Comment (BeniaminDrasovean) [inline arwen/helpers.go:309]: Missing comment.

File: arwen/helpers.go
Comment (BeniaminDrasovean) [inline arwen/helpers.go:316]: Missing comment.

**Review by BeniaminDrasovean — CHANGES_REQUESTED** [review]

File: arwen/common.go
Comment (camilbancioiu) [reply arwen/common.go:14]: Opened this PR from master, to be merged into master: https://github.com/ElrondNetwork/arwen-wasm-vm/pull/218. It adds storage protection in a better way, and a unit test.

File: arwen/contexts/async.go
Comment (camilbancioiu) [reply arwen/contexts/async.go:125]: Added comment.

File: arwen/contexts/async.go
Comment (camilbancioiu) [reply arwen/contexts/async.go:116]: Added.

File: arwen/contexts/validator.go
Comment (camilbancioiu) [reply arwen/contexts/validator.go:113]: Good catch. Fixed.

File: arwen/errors.go
Comment (camilbancioiu) [reply arwen/errors.go:85]: Will be added after merging the linting fixes made on master, to avoid conflicts. But will not merge the linting fixes, to keep the PR readable.

File: integrationTests/json/vm_test.go
Comment (camilbancioiu) [reply integrationTests/json/vm_test.go:338]: Haven't tried yet, but the contracts will likely need minor changes. This is the next step, though.

File: mock/vmHostMock.go
Comment (camilbancioiu) [reply mock/vmHostMock.go:165]: Yes, better.

File: arwen/contexts/async.go
Comment (camilbancioiu) [reply arwen/contexts/async.go:679]: It's the cost of passing the returned data from the destination call into the arguments of the callback. It must match the cost of the `Data` field of a SCR would have, for cross-shard calls.

File: arwen/contexts/async.go
Comment (camilbancioiu) [reply arwen/contexts/async.go:259]: `findCall` does not return any other error.

Simplified with `if err != nil { return false }`, though, because it's equivalent.

File: arwen/asyncCallGroup.go
Comment (camilbancioiu) [reply arwen/asyncCallGroup.go:56]: Fixed.

File: arwen/asyncCallGroup.go
Comment (camilbancioiu) [reply arwen/asyncCallGroup.go]: The `AsyncCallHandler` interface will be removed soon. Would it be better to keep it? The code that uses it doesn't cross packages, it's likely okay to just use the `struct`.

File: arwen/common.go
Comment (camilbancioiu) [reply arwen/common.go:12]: Fixed.

File: arwen/common.go
Comment (camilbancioiu) [reply arwen/common.go]: Fixed.

File: arwen/common.go
Comment (camilbancioiu) [reply arwen/common.go]: Fixed.

File: arwen/common.go
Comment (camilbancioiu) [reply arwen/common.go]: Fixed.

File: arwen/contexts/async.go
Comment (camilbancioiu) [reply arwen/contexts/async.go:117]: Added,

File: arwen/contexts/async.go
Comment (camilbancioiu) [reply arwen/contexts/async.go:126]: Added.

File: arwen/contexts/async.go
Comment (camilbancioiu) [reply arwen/contexts/async.go]: Fixed.

File: arwen/contexts/async.go
Comment (camilbancioiu) [reply arwen/contexts/async.go]: Fixed.

File: arwen/contexts/async_test.go
Comment (camilbancioiu) [reply arwen/contexts/async_test.go:199]: Fixed.

File: arwen/helpers.go
Comment (camilbancioiu) [reply arwen/helpers.go:276]: Added.

File: arwen/helpers.go
Comment (camilbancioiu) [reply arwen/helpers.go:291]: Added.

File: arwen/helpers.go
Comment (camilbancioiu) [reply arwen/helpers.go:304]: Added.

File: arwen/helpers.go
Comment (camilbancioiu) [reply arwen/helpers.go:309]: Added.

File: arwen/helpers.go
Comment (camilbancioiu) [reply arwen/helpers.go:316]: Added.

**Comment by camilbancioiu** [comment]
Conflicts will be fixed after review.

File: arwen/asyncCallGroup.go
Comment (BeniaminDrasovean) [reply arwen/asyncCallGroup.go]: For the sake of better code abstractization, and for future development, I think it is better to keep it.

**Review by BeniaminDrasovean — APPROVED** [review]

File: arwen/contexts/async.go
Comment (sasurobert) [reply arwen/contexts/async.go:702]: it could be simply 2xdataLen :D

**Review by sasurobert — APPROVED** [review]

**Pull request merged**
